File: apps/chat/libs/chat-storage/src/main/java/com/example/chat/storage/mapper/ChannelMetadataMapper.java
package com.example.chat.storage.mapper;

import org.springframework.stereotype.Component;

import com.example.chat.domain.channel.ChannelId;
import com.example.chat.domain.channel.metadata.ChannelMetadata;
import com.example.chat.domain.channel.metadata.ChannelMetadataId;
import com.example.chat.domain.message.MessageId;
import com.example.chat.domain.user.UserId;
import com.example.chat.storage.entity.ChatChannelMetadataEntity;

/**
 * ChannelMetadata Domain ↔ ChatChannelMetadataEntity 변환
 */
@Component
public class ChannelMetadataMapper {

    /**
     * Domain → Entity 변환
     */
    public ChatChannelMetadataEntity toEntity(ChannelMetadata metadata) {
        return ChatChannelMetadataEntity.builder()
                .id(metadata.getId().value())
                .channelId(metadata.getChannelId().value())
                .userId(metadata.getUserId().value())
                .notificationEnabled(metadata.isNotificationEnabled())
                .favorite(metadata.isFavorite())
                .pinned(metadata.isPinned())
                .lastReadMessageId(
                        metadata.getLastReadMessageId() != null ? metadata.getLastReadMessageId().value() : null)
                .lastReadAt(metadata.getLastReadAt())
                .unreadCount(metadata.getUnreadCount())
                .lastActivityAt(metadata.getLastActivityAt())
                .createdAt(metadata.getCreatedAt())
                .updatedAt(metadata.getUpdatedAt())
                .build();
    }

    /**
     * Entity → Domain 변환
     */
    public ChannelMetadata toDomain(ChatChannelMetadataEntity entity) {
        return ChannelMetadata.fromStorage(
                ChannelMetadataId.of(entity.getId()),
                ChannelId.of(entity.getChannelId()),
                UserId.of(entity.getUserId()),
                entity.getCreatedAt(),
                entity.isNotificationEnabled(),
                entity.isFavorite(),
                entity.isPinned(),
                entity.getLastReadMessageId() != null ? MessageId.of(entity.getLastReadMessageId()) : null,
                entity.getLastReadAt(),
                entity.getUnreadCount(),
                entity.getLastActivityAt() != null ? entity.getLastActivityAt() : entity.getCreatedAt(),
                entity.getUpdatedAt());
    }
}

File: apps/chat/libs/chat-storage/src/main/kotlin/com/example/chat/storage/mapper/ChannelMetadataMapper.kt
package com.example.chat.storage.mapper

import com.example.chat.domain.channel.ChannelId
import com.example.chat.domain.channel.metadata.ChannelMetadata
import com.example.chat.domain.channel.metadata.ChannelMetadataId
import com.example.chat.domain.message.MessageId
import com.example.chat.domain.user.UserId
import com.example.chat.storage.entity.ChatChannelMetadataEntity
import org.springframework.stereotype.Component

/**
 * ChannelMetadata Domain ↔ ChatChannelMetadataEntity 변환
 */
@Component
class ChannelMetadataMapper {

	/**
	 * Domain → Entity 변환
	 */
	fun toEntity(metadata: ChannelMetadata): ChatChannelMetadataEntity {
		return ChatChannelMetadataEntity(
			id = metadata.id.value,
			channelId = metadata.channelId.value,
			userId = metadata.userId.value,
			notificationEnabled = metadata.notificationEnabled,
			favorite = metadata.favorite,
			pinned = metadata.pinned,
			lastReadMessageId = metadata.lastReadMessageId?.value,
			lastReadAt = metadata.lastReadAt,
			unreadCount = metadata.unreadCount,
			lastActivityAt = metadata.lastActivityAt,
			createdAt = metadata.createdAt,
			updatedAt = metadata.updatedAt
		)
	}

	/**
	 * Entity → Domain 변환
	 */
	fun toDomain(entity: ChatChannelMetadataEntity): ChannelMetadata {
		return ChannelMetadata(
			id = ChannelMetadataId.of(entity.id),
			channelId = ChannelId.of(entity.channelId),
			userId = UserId.of(entity.userId),
			notificationEnabled = entity.notificationEnabled,
			favorite = entity.favorite,
			pinned = entity.pinned,
			lastReadMessageId = entity.lastReadMessageId?.let { MessageId.of(it) },
			lastReadAt = entity.lastReadAt,
			unreadCount = entity.unreadCount,
			lastActivityAt = entity.lastActivityAt ?: entity.createdAt,
			createdAt = entity.createdAt,
			updatedAt = entity.updatedAt
		)
	}
}

/**
 * Extension Functions for more idiomatic Kotlin
 */
fun ChannelMetadata.toEntity(): ChatChannelMetadataEntity {
	return ChatChannelMetadataEntity(
		id = id.value,
		channelId = channelId.value,
		userId = userId.value,
		notificationEnabled = notificationEnabled,
		favorite = favorite,
		pinned = pinned,
		lastReadMessageId = lastReadMessageId?.value,
		lastReadAt = lastReadAt,
		unreadCount = unreadCount,
		lastActivityAt = lastActivityAt,
		createdAt = createdAt,
		updatedAt = updatedAt
	)
}

fun ChatChannelMetadataEntity.toDomain(): ChannelMetadata {
	return ChannelMetadata(
		id = ChannelMetadataId.of(id),
		channelId = ChannelId.of(channelId),
		userId = UserId.of(userId),
		notificationEnabled = notificationEnabled,
		favorite = favorite,
		pinned = pinned,
		lastReadMessageId = lastReadMessageId?.let { MessageId.of(it) },
		lastReadAt = lastReadAt,
		unreadCount = unreadCount,
		lastActivityAt = lastActivityAt ?: createdAt,
		createdAt = createdAt,
		updatedAt = updatedAt
	)
}
